---
title: 'Package walkthrough: sleuth'
author: ''
date: '2018-04-28'
slug: package-walkthrough-sleuth
categories:
  - R
  - package
tags:
  - bioinfo_tools
---

<script src="/rmarkdown-libs/kePrint/kePrint.js"></script>


<div id="kallisto-sleuth-workflow-in-dealing-with-sequencing-data" class="section level1">
<h1>kallisto-sleuth workflow in dealing with sequencing data</h1>
<p>Needless to say, <a href="https://pachterlab.github.io/kallisto/">kallisto</a>-<a href="https://pachterlab.github.io/sleuth/">sleuth</a> workflow opens up a new era in dealing with sequencing data; it dramatically scales down the computational complexity to the point which sequencing can be handled by a PC. I find <a href="http://tinyheero.github.io/2015/09/02/pseudoalignments-kallisto.html">this article is a good startpoint to understand the mechanism of kallisto</a></p>
<p>As the downstream of kallisto, sleuth has been optimised to deal with kallisto output. An attactive concept from sleuth is <a href="https://www.nature.com/articles/nmeth.4324">“variance decoupling”</a>, to “decouple biological variance from inferential variance.” Inferential variance, according to online method, consists of “<em>errors in mesurement</em> and subsequent inference”, which can be obtained through bootstrapping from kallisto; inferential variance is eliminated, or decoupled away before shrinkage (and later added back when estimating the fixed-effect parameter). This articles walks through R code that realises this decoupling process.</p>
</div>
<div id="dataset-used-and-method-to-open-up-packages" class="section level1">
<h1>Dataset used and method to “open up packages”</h1>
<div id="kallisto-pre-processing" class="section level4">
<h4>Kallisto pre-processing</h4>
<p>We obtained the data as descripted in <a href="http://sepsis-omics.github.io/tutorials/modules/kallisto/">“Get data” section from “Differential gene expression using Kallisto and Degust”</a>. We processed using the following bash code.</p>
<pre class="bash"><code>cd /your_dirs_to_Ecoli_kallisto_files

kallisto index -i transcripts.idx Ecoli_transcripts.fasta

kallisto quant -b 100 -i transcripts.idx -o LB1 --single -l 500 -s 50 LB1.fastq.gz
kallisto quant -b 100 -i transcripts.idx -o LB2 --single -l 500 -s 50 LB2.fastq.gz
kallisto quant -b 100 -i transcripts.idx -o LB3 --single -l 500 -s 50 LB3.fastq.gz
kallisto quant -b 100 -i transcripts.idx -o MG1 --single -l 500 -s 50 MG1.fastq.gz
kallisto quant -b 100 -i transcripts.idx -o MG2 --single -l 500 -s 50 MG2.fastq.gz
kallisto quant -b 100 -i transcripts.idx -o MG3 --single -l 500 -s 50 MG3.fastq.gz</code></pre>
</div>
<div id="method-to-open-up-sleuth-packages." class="section level4">
<h4>Method to open up sleuth packages.</h4>
<p>The aforementioned bash script yielded a dataset that is a typical “3 treatment Vs 3 control” experimental design. According to <a href="https://pachterlab.github.io/sleuth_walkthroughs/trapnell/analysis.html">the official manual</a>, dealing with it requires 4 lines with 3 functions</p>
<ul>
<li>one <code>sleuth_prep</code> performs filtering (default as &lt;5 count in 47% samples) and sample normalisation.</li>
<li>two <code>sleuth_fit</code> perform, on full model (with treatment) and null model (without treatment) respectively, inferential variance decoupling and using the variance left (biological variance) to perform odinary lease sq estimation of coefficient.</li>
<li>one <code>sleuth_lrt</code> perform likelihood ratio test between full and null model to see which model is likely to account for reality.</li>
</ul>
<p>There are vaious intermediates that are wrapped wihin these functions. To understand these function, we have to find a way to examine the input/output of each intermediate steps. We modified these three functions which are summarised as follows,</p>
<ul>
<li>adding <code>sleuth:::</code> to seek internal functions from sleuth library. By doing this, we do not have to touch the rest of packages.</li>
<li><p>assign the intermediate outputs, which we might be interested in, to global variables using <code>&lt;&lt;-</code>. A three-letter code was prefixed to the variables’ original names; these are</p>
<ul>
<li>“PPP_” for variable in <code>sleuth_prep</code> function.</li>
<li>“FFF_” for variable in <code>sleuth_fit</code> on full model.</li>
<li>“RRR_” for varaible in <code>sleuth_fit</code> on reduced model.</li>
<li>“LLL_” for variable in <code>sleuth_lrt</code> function.</li>
</ul></li>
</ul>
<p>The modified code is stored <a href="/sleuth/sleuth_internal_structure.R">“sleuth_internal_structure.R”</a> and will be sourced in and run as a replacemnt of normal code. As a result, all the intermediate outputs will be “recorded”.</p>
<pre class="r"><code>## setwd(paste0(getwd(), &quot;/content/package&quot;)) ## needed to run this code chunk alone instead of Rmd compilation. 
if (packageVersion(&quot;sleuth&quot;) != &quot;0.29.0&quot;){
    devtools::install_github(&quot;pachterlab/sleuth@v0.29.0&quot;)  
}

x &lt;- c(&quot;tidyverse&quot;,&quot;sleuth&quot;, &quot;kableExtra&quot;)
x &lt;- lapply(x, require, character.only = TRUE)
source(&quot;../../static/sleuth/sleuth_internal_structure.R&quot;) 

## read the experiment design matrix
t &lt;- read.table(&quot;../../static/sleuth/tbl.txt&quot;, header = T)
t$path &lt;- as.character(t$path)

obj &lt;&lt;- test &lt;- PPP_sleuth_prep(t, extra_bootstrap_summary = TRUE, num_cores = 1)
test &lt;- FFF_sleuth_fit(test, ~condition, &#39;full&#39;)
test &lt;- RRR_sleuth_fit(test, ~1, &#39;reduced&#39;)
test &lt;- LLL_sleuth_lrt(test, &#39;reduced&#39;, &#39;full&#39;)</code></pre>
</div>
</div>
<div id="the-walkthrough" class="section level1">
<h1>The walkthrough</h1>
<pre class="r"><code>options(knitr.table.format = &quot;html&quot;, kableExtra.auto_format = TRUE, kableExtra.font_size = 8)
fs &lt;- 8</code></pre>
<p><code>sleuth_prep</code> output is a list of 19; the observed count and bootstrapping information is stored in <code>bs_summary</code>.</p>
<pre class="r"><code>str(obj$bs_summary)</code></pre>
<pre><code>## List of 2
##  $ obs_counts: num [1:1481, 1:6] 4.62 3.13 3.53 2.46 5.6 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:1481] &quot;b0002&quot; &quot;b0003&quot; &quot;b0004&quot; &quot;b0006&quot; ...
##   .. ..$ : chr [1:6] &quot;SRR493366&quot; &quot;SRR493367&quot; &quot;SRR493368&quot; &quot;SRR493369&quot; ...
##  $ sigma_q_sq: Named num [1:1481] 0.02105 0.07362 0.03354 0.049 0.00279 ...
##   ..- attr(*, &quot;names&quot;)= chr [1:1481] &quot;b0002&quot; &quot;b0003&quot; &quot;b0004&quot; &quot;b0006&quot; ...</code></pre>
<p>Within <code>sleuth_fit</code>, the function <code>sleuth:::me_model</code> wrapped in <code>sleuth:::me_model_by_row</code> elimiates inferential variance from total variance. Inferrential variance (variable <code>sigma_q_sq</code>) is obtained from <code>rowMean</code> of the <code>bs_summary</code>.</p>
<pre class="r"><code>## modified from me_model and me_model_by_row
me_model_by_row &lt;- lapply(1:nrow(bs_summary$obs_counts), function(i) {
  function(X, bs_summary$obs_counts[i, ], bs_summary$sigma_q_sq[i]) {
    n &lt;- nrow(X)
    degrees_free &lt;- n - ncol(X)
    ols_fit &lt;- lm.fit(X, y)
    rss &lt;- sum(ols_fit$residuals^2) # residual sum sq
    
    ## (raw) biological_var = total_var - inferential_var
    sigma_sq &lt;- rss/(degrees_free) - sigma_q_sq 
    
    ## collectin result 
    mean_obs &lt;- mean(y)
    var_obs &lt;- var(y)
    list(ols_fit = ols_fit, b1 = ols_fit$coefficients[2], rss = rss, 
         sigma_sq = sigma_sq, sigma_q_sq = sigma_q_sq, mean_obs = mean_obs, 
         var_obs = var_obs)
    }
})</code></pre>
<p>Subsequently, <code>sigma_sq_pmax</code> is obtained from <code>sigma_sq</code> of each gene by <code>sigma_sq_pmax = pmax(sigma_sq,0)</code>. This is to prevent negative variance of count.</p>
<pre class="r"><code>head(FFF_mes_df) %&gt;%
  kable(.) %&gt;%
  kable_styling(font_size = fs) </code></pre>
<table class="table" style="font-size: 8px; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
rss
</th>
<th style="text-align:right;">
sigma_sq
</th>
<th style="text-align:right;">
sigma_q_sq
</th>
<th style="text-align:right;">
mean_obs
</th>
<th style="text-align:right;">
var_obs
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:right;">
sigma_sq_pmax
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.1104198
</td>
<td style="text-align:right;">
0.0065574
</td>
<td style="text-align:right;">
0.0210476
</td>
<td style="text-align:right;">
4.084207
</td>
<td style="text-align:right;">
0.2834129
</td>
<td style="text-align:left;">
b0002
</td>
<td style="text-align:right;">
0.0065574
</td>
</tr>
<tr>
<td style="text-align:right;">
0.6804381
</td>
<td style="text-align:right;">
0.0964892
</td>
<td style="text-align:right;">
0.0736203
</td>
<td style="text-align:right;">
2.803340
</td>
<td style="text-align:right;">
0.2939586
</td>
<td style="text-align:left;">
b0003
</td>
<td style="text-align:right;">
0.0964892
</td>
</tr>
<tr>
<td style="text-align:right;">
0.1362586
</td>
<td style="text-align:right;">
0.0005256
</td>
<td style="text-align:right;">
0.0335390
</td>
<td style="text-align:right;">
3.376731
</td>
<td style="text-align:right;">
0.0424544
</td>
<td style="text-align:left;">
b0004
</td>
<td style="text-align:right;">
0.0005256
</td>
</tr>
<tr>
<td style="text-align:right;">
0.2719853
</td>
<td style="text-align:right;">
0.0189926
</td>
<td style="text-align:right;">
0.0490038
</td>
<td style="text-align:right;">
3.071661
</td>
<td style="text-align:right;">
0.2606999
</td>
<td style="text-align:left;">
b0006
</td>
<td style="text-align:right;">
0.0189926
</td>
</tr>
<tr>
<td style="text-align:right;">
0.0323607
</td>
<td style="text-align:right;">
0.0052969
</td>
<td style="text-align:right;">
0.0027933
</td>
<td style="text-align:right;">
5.873750
</td>
<td style="text-align:right;">
0.0943577
</td>
<td style="text-align:left;">
b0008
</td>
<td style="text-align:right;">
0.0052969
</td>
</tr>
<tr>
<td style="text-align:right;">
0.0722622
</td>
<td style="text-align:right;">
0.0166392
</td>
<td style="text-align:right;">
0.0014264
</td>
<td style="text-align:right;">
6.367659
</td>
<td style="text-align:right;">
0.1211020
</td>
<td style="text-align:left;">
b0014
</td>
<td style="text-align:right;">
0.0166392
</td>
</tr>
</tbody>
</table>
<p><code>sleuth:::sliding_window_grouping</code> (uses <code>ecdf</code>) and <code>sleuth:::shrink_df</code> perform loess prediction using biological variance.</p>
<pre class="r"><code>head(FFF_swg) %&gt;%
  kable(., &quot;html&quot;) %&gt;%
  kable_styling(font_size = fs)</code></pre>
<table class="table" style="font-size: 8px; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
rss
</th>
<th style="text-align:right;">
sigma_sq
</th>
<th style="text-align:right;">
sigma_q_sq
</th>
<th style="text-align:right;">
mean_obs
</th>
<th style="text-align:right;">
var_obs
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:right;">
sigma_sq_pmax
</th>
<th style="text-align:left;">
iqr
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1.0236176
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:right;">
0.1394486
</td>
<td style="text-align:right;">
1.1483599
</td>
<td style="text-align:right;">
2.748282
</td>
<td style="text-align:left;">
b0071
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
1.6284312
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:right;">
0.2882827
</td>
<td style="text-align:right;">
1.1150755
</td>
<td style="text-align:right;">
1.040309
</td>
<td style="text-align:left;">
b0128
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
3.2163447
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:right;">
0.2621964
</td>
<td style="text-align:right;">
1.2032913
</td>
<td style="text-align:right;">
1.017009
</td>
<td style="text-align:left;">
b0325
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
1.3087485
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:right;">
0.2407820
</td>
<td style="text-align:right;">
1.1389325
</td>
<td style="text-align:right;">
1.376002
</td>
<td style="text-align:left;">
b0859
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7502072
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:right;">
0.1837905
</td>
<td style="text-align:right;">
0.6699952
</td>
<td style="text-align:right;">
1.378772
</td>
<td style="text-align:left;">
b0897
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:right;">
3.7595103
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:right;">
0.1955008
</td>
<td style="text-align:right;">
1.1707549
</td>
<td style="text-align:right;">
1.019893
</td>
<td style="text-align:left;">
b1199
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>head(FFF_shrink1) %&gt;%
  kable(., &quot;html&quot;) %&gt;%
  kable_styling(font_size = fs)</code></pre>
<table class="table" style="font-size: 8px; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
rss
</th>
<th style="text-align:right;">
sigma_sq
</th>
<th style="text-align:right;">
sigma_q_sq
</th>
<th style="text-align:right;">
mean_obs
</th>
<th style="text-align:right;">
var_obs
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:right;">
sigma_sq_pmax
</th>
<th style="text-align:left;">
iqr
</th>
<th style="text-align:right;">
shrink
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1.0236176
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:right;">
0.1394486
</td>
<td style="text-align:right;">
1.1483599
</td>
<td style="text-align:right;">
2.748282
</td>
<td style="text-align:left;">
b0071
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.7539531
</td>
</tr>
<tr>
<td style="text-align:right;">
1.6284312
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:right;">
0.2882827
</td>
<td style="text-align:right;">
1.1150755
</td>
<td style="text-align:right;">
1.040309
</td>
<td style="text-align:left;">
b0128
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.7663570
</td>
</tr>
<tr>
<td style="text-align:right;">
3.2163447
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:right;">
0.2621964
</td>
<td style="text-align:right;">
1.2032913
</td>
<td style="text-align:right;">
1.017009
</td>
<td style="text-align:left;">
b0325
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.7339655
</td>
</tr>
<tr>
<td style="text-align:right;">
1.3087485
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:right;">
0.2407820
</td>
<td style="text-align:right;">
1.1389325
</td>
<td style="text-align:right;">
1.376002
</td>
<td style="text-align:left;">
b0859
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.7574440
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7502072
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:right;">
0.1837905
</td>
<td style="text-align:right;">
0.6699952
</td>
<td style="text-align:right;">
1.378772
</td>
<td style="text-align:left;">
b0897
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:right;">
3.7595103
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:right;">
0.1955008
</td>
<td style="text-align:right;">
1.1707549
</td>
<td style="text-align:right;">
1.019893
</td>
<td style="text-align:left;">
b1199
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.7457314
</td>
</tr>
</tbody>
</table>
<p>A power of 4 is performed on <code>shrink</code> variable to produce <code>smooth_sigma_sq</code>.</p>
<pre class="r"><code>head(FFF_shrink_select) %&gt;%
  kable(., &quot;html&quot;) %&gt;%
  kable_styling(font_size = fs)</code></pre>
<table class="table" style="font-size: 8px; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
rss
</th>
<th style="text-align:right;">
sigma_sq
</th>
<th style="text-align:right;">
sigma_q_sq
</th>
<th style="text-align:right;">
mean_obs
</th>
<th style="text-align:right;">
var_obs
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:right;">
sigma_sq_pmax
</th>
<th style="text-align:left;">
iqr
</th>
<th style="text-align:right;">
smooth_sigma_sq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1.0236176
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:right;">
0.1394486
</td>
<td style="text-align:right;">
1.1483599
</td>
<td style="text-align:right;">
2.748282
</td>
<td style="text-align:left;">
b0071
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.3231301
</td>
</tr>
<tr>
<td style="text-align:right;">
1.6284312
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:right;">
0.2882827
</td>
<td style="text-align:right;">
1.1150755
</td>
<td style="text-align:right;">
1.040309
</td>
<td style="text-align:left;">
b0128
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.3449249
</td>
</tr>
<tr>
<td style="text-align:right;">
3.2163447
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:right;">
0.2621964
</td>
<td style="text-align:right;">
1.2032913
</td>
<td style="text-align:right;">
1.017009
</td>
<td style="text-align:left;">
b0325
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.2902034
</td>
</tr>
<tr>
<td style="text-align:right;">
1.3087485
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:right;">
0.2407820
</td>
<td style="text-align:right;">
1.1389325
</td>
<td style="text-align:right;">
1.376002
</td>
<td style="text-align:left;">
b0859
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.3291563
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7502072
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:right;">
0.1837905
</td>
<td style="text-align:right;">
0.6699952
</td>
<td style="text-align:right;">
1.378772
</td>
<td style="text-align:left;">
b0897
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:right;">
3.7595103
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:right;">
0.1955008
</td>
<td style="text-align:right;">
1.1707549
</td>
<td style="text-align:right;">
1.019893
</td>
<td style="text-align:left;">
b1199
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.3092642
</td>
</tr>
</tbody>
</table>
<p>The maximun of variance was obtained between raw and shrunken estimate.</p>
<pre class="r"><code>head(FFF_shrink_mutate) %&gt;%
  kable(., &quot;html&quot;) %&gt;%
  kable_styling(font_size = fs)</code></pre>
<table class="table" style="font-size: 8px; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;">
rss
</th>
<th style="text-align:right;">
sigma_sq
</th>
<th style="text-align:right;">
sigma_q_sq
</th>
<th style="text-align:right;">
mean_obs
</th>
<th style="text-align:right;">
var_obs
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:right;">
sigma_sq_pmax
</th>
<th style="text-align:left;">
iqr
</th>
<th style="text-align:right;">
smooth_sigma_sq
</th>
<th style="text-align:right;">
smooth_sigma_sq_pmax
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1.0236176
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:right;">
0.1394486
</td>
<td style="text-align:right;">
1.1483599
</td>
<td style="text-align:right;">
2.748282
</td>
<td style="text-align:left;">
b0071
</td>
<td style="text-align:right;">
0.1164558
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.3231301
</td>
<td style="text-align:right;">
0.3231301
</td>
</tr>
<tr>
<td style="text-align:right;">
1.6284312
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:right;">
0.2882827
</td>
<td style="text-align:right;">
1.1150755
</td>
<td style="text-align:right;">
1.040309
</td>
<td style="text-align:left;">
b0128
</td>
<td style="text-align:right;">
0.1188251
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.3449249
</td>
<td style="text-align:right;">
0.3449249
</td>
</tr>
<tr>
<td style="text-align:right;">
3.2163447
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:right;">
0.2621964
</td>
<td style="text-align:right;">
1.2032913
</td>
<td style="text-align:right;">
1.017009
</td>
<td style="text-align:left;">
b0325
</td>
<td style="text-align:right;">
0.5418898
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.2902034
</td>
<td style="text-align:right;">
0.5418898
</td>
</tr>
<tr>
<td style="text-align:right;">
1.3087485
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:right;">
0.2407820
</td>
<td style="text-align:right;">
1.1389325
</td>
<td style="text-align:right;">
1.376002
</td>
<td style="text-align:left;">
b0859
</td>
<td style="text-align:right;">
0.0864052
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.3291563
</td>
<td style="text-align:right;">
0.3291563
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7502072
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:right;">
0.1837905
</td>
<td style="text-align:right;">
0.6699952
</td>
<td style="text-align:right;">
1.378772
</td>
<td style="text-align:left;">
b0897
</td>
<td style="text-align:right;">
0.0037614
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:right;">
3.7595103
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:right;">
0.1955008
</td>
<td style="text-align:right;">
1.1707549
</td>
<td style="text-align:right;">
1.019893
</td>
<td style="text-align:left;">
b1199
</td>
<td style="text-align:right;">
0.7443768
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.3092642
</td>
<td style="text-align:right;">
0.7443768
</td>
</tr>
</tbody>
</table>
<p>The fixed-effect beta is estimated with <code>sleuth:::covar_beta</code> using total vairance, which is the shrunken biological variance <code>smooth_sigma_sq_pmax</code> plus inference variance <code>sigma_q_sq</code>.</p>
<p>In the later part, the likelihood of each model is calculated by</p>
<pre class="r"><code>all_likelihood &lt;- sapply(nrow, function(i) {
    cur_mu &lt;- fitted.values
    obs &lt;- residuals + cur_mu
    cur_summary &lt;- obj$fits[[which_model]]$summary
    cur_var &lt;- smooth_sigma_sq_pmax + sigma_q_sq
    sum(dnorm(obs, mean = cur_mu, sd = sqrt(cur_var),log = TRUE)) ## normal distribution!!
})</code></pre>
<p>… with test stats as
<span class="math display">\[2 \times (l_{full}- l_{null})\]</span>
P-value is from <code>pchisq</code> and q-value from <code>p.adjust(pval, method = &quot;BH&quot;)</code>.</p>
</div>
